Convert module code from JS to TypeScript

mangs opened the pull request:
**Changes:**

- Version bump to `2.0.0`
- Convert code to TypeScript
- Convert code building from Rollup to Vite v4
- Convert unit testing from Jest to Vitest

mangs commented on src/RollbarClientSubmitter.ts:
```diff
@@ -12,31 +21,41 @@ const configurationDefaults = {
-  if (targetValue === null || typeof targetValue !== 'object' || Array.isArray(targetValue)) {
-    return targetValue;
-  }
+// function buildObjectDeepSorted(targetValue: any) {
```
This is the implementation from earlier today. I modified and rearranged it (see below) to simplify the logic somewhat and added generic type handling which overall has reduced the number of errors, but there is still one left.

mangs commented:
I'm not sure why the ESLint error count shot through the roof since adding Vite build support. Perhaps when I remove Jest and add Vitest things will calm down.

mangs commented:
@jduthon Are you working this week? I'm still working on this, I'm done converting this to Vite and Vitest, so all that's left is cleaning up the remaining few `tsc` and ESLint errors. The book I got is pretty helpful, but if you have any ideas I'd appreciate any tips. Happy holidays.

jduthon commented:
@mangs I have opened a PR some time back on your fork of the repo to address the issues we had discussed [here](https://github.com/mangs/rollbar-client.js/pull/1)

mangs commented:
@jduthon not sure how I missed that, thank you! I'll take a look.

mangs commented on package.json:
```diff
@@ -16,63 +16,50 @@
-    "update:package-versions": "./scripts/npm/updatePackageVersions.mjs --ignore-packages=rollup && npm run --silent reinstall",
+    "test": "npm-run-all --silent validate:environment test:unit",
+    "test:unit": "vitest --config config/vite/vite.config.mts --coverage",
+    "test:unit:ui": "npm run --silent test:unit -- --ui",
```
This is a really nice feature, worth trying when you have time. It provides a nice UI for feedback for your running tests.

mangs commented on src/RollbarClient.mts:
```diff
@@ -0,0 +1,50 @@
+// Internal Imports
```
I did the same `git mv` command for all these file renames, but unfortunately some of them show up as deletes then creates :(

mangs commented on src/RollbarClient.test.mts:
```diff
@@ -1,23 +1,28 @@
-}));
+vi.mock('./RollbarClientSubmitter', () => {
+  const RollbarClientSubmitter = vi.fn();
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- TODO: determine how to type prototype as non-any
```
I was stuck for a long time on this where it was complaining about `.prototype` being `any`; I eventually decided to move on. I don't think this is a huge deal, but if you have any ideas I'll certainly consider them.

mangs replied:
I thought so too, but the Vitest documentation specifically advises to uses prototype in this way to mock classes (see "Mock exported class implementation"): https://vitest.dev/guide/mocking.html#:~:text=Mock%20exported%20class%20implementation

Using `class` causes the same error too confusingly

jduthon replied:
Huh interesting.

In this same docs block they also have:

> Example with vi.mock and return value:

> ts
import { SomeClass } from 'some-path'
vi.mock('some-path', () => {
  const SomeClass = vi.fn(() => ({
    someMethod: vi.fn()
  }))
  return { SomeClass }
})
// SomeClass.mock.returns will have returned object

So I guess we could try that instead.

Though if the class gives the same error as the prototype, I think it means the problem is not with `prototype` but with the type of `RollbarClientSubmitter` within this snippet, which would make sense as it is never defined what type `RollbarClientSubmitter` should be here.

So I think you probably need something liket his if `vi.fn` is a generic:
`const RollbarClientSubmitter = vi.fn<RollbarClientSubmitter>();` (where the `RollbarClientSubmitter` between chevrons should be the original one as we want to use it as a type)
or if `vi.fn` isn't a generic:
`const RollbarClientSubmitter = vi.fn() as RollbarClientSubmitter;`

I think that should rid the code of the warning.

jduthon replied:
Ugh classes just suck 😅 

After playing a little bit with it, the best I can do is:
```typescript
  const MockRollbarClientSubmitter = vi.fn(() => ({
    report: vi.fn(),
  } as Partial<RollbarClientSubmitter>));
  return { RollbarClientSubmitter: MockRollbarClientSubmitter };
```

NB: I need the `Partial` as the mock doesn't implement all the functions exposed by `RollbarClientSubmitter`, while it's ok, it's also risky as technically nothing would prevent `RollbarClient` to use more things from `RollbarClientSubmitter` as they aren't private.
I think, if internals were marked as private, then the Partial could be dropped, but somehow you need to extract the keys instead of using the type directly like so:

```typescript
  const MockRollbarClientSubmitter = vi.fn(() => ({
    report: vi.fn(),
  } as {[K in keyof RollbarClientSubmitter]: RollbarClientSubmitter[K]}));
  return { RollbarClientSubmitter: MockRollbarClientSubmitter };
 ```

mangs replied:
Haha. Classes can be confounding for sure.

Thank you for giving this a shot! My concern here is you have not mimic'd a class here, rather when instantiating a new object a new `report` function object will be created whereas attaching it to `prototype` will reuse the same function object for every instantiation. I'd like to stick to a class-like data structure here as closely as possible. What are your thoughts on this?

Good suggestion about private class fields, but unfortunately Safari support isn't quite far enough along relative to our userbase for me to feel comfortable using it yet: https://caniuse.com/mdn-javascript_classes_private_class_fields

If I were to use private class fields, I would have to move some of my tests into the code files themselves a.k.a. "in-source testing" (https://vitest.dev/guide/in-source.html, Rust also does this which is where they got the idea I think) which I need to think more about. Basically anything inaccessible from an external module scope would have to be tested that way. This is something Jest does not provide that I plan to play with in the future.

jduthon replied:
> Thank you for giving this a shot! My concern here is you have not mimic'd a class here, rather when instantiating a new object a new report function object will be created whereas attaching it to prototype will reuse the same function object for every instantiation. I'd like to stick to a class-like data structure here as closely as possible. What are your thoughts on this?

You could get one function handling all also with the structure above
```typescript
const MockRollbarClientSubmitterPrototype = {
  report: vi.fn(),
};
const MockRollbarClientSubmitter = vi.fn(() => MockRollbarClientSubmitterPrototype);
return { RollbarClientSubmitter: MockRollbarClientSubmitter };
```
but yeah there are probably other differences between class VS object, though chances are you also already have some differences in the current code (because `vi.fn()` isn't a class).

For private class fields, you could use them in TypeScript but transpile them (that's what I had in mind personally)

But just in case none of this is a blocker for me, ultimately we are typing things in a test, it can be nice to get the types right as it'll make writing tests easier, but it's clearly not the most important.

mangs replied:
Thank you for your useful feedback here, but I'd like to leave this as-is as a reminder to iterate on this a bit in the future. I will keep your comments in mind for sure.

For private class fields, because I'm not using `tsc` to build the code, I can't specify compatibility targets. I'd have to look to see what `esbuild`'s built-in transpilation target is. I know it's configurable, but it is not compatible with `browserslist` which is something I'd like to examine more in the future. I'll keep `#private_field`s in mind (I assume you were not talking about `private field` which is not a runtime private field).

jduthon replied:
I was thinking of `private field`, while I agree it's useless for JavaScript, it isn't useless for TypeScript ;)

mangs replied:
`#private` gives you the benefits in both places which was why I was interested in using it

mangs commented on src/RollbarClientSubmitter.test.mts:
```diff
@@ -411,7 +414,7 @@ describe(`Class: ${RollbarClientSubmitter.name}`, () => {
           submitter.report('error', testMessage);
 
-          const actualPayload = JSON.parse(navigator.sendBeacon.mock.calls[0][1]);
+          const actualPayload = JSON.parse(navigator.sendBeacon.mock.calls[0][1]) as IPayload;
```
This is the only line **in the code** with a `tsc` error remaining. I can't get `tsc` or `eslint` to treat `sendBeacon` as a mocked function, so I'm a little stumped here.

mangs replied:
I ended up ignoring these errors due to not being able to get to the bottom of this problem in a timely manner. As you can see, I've added a `TODO` comment to keep an eye on this so when this issue is fixed in the future (or maybe I'm doing something wrong?) I'll remember to address it.

mangs replied:
Unfortunately these are many of the things I've tried. I had to add the `@ts-ignore` to make the `tsc` complaints go away. I tried both of your suggestions and neither worked unfortunately. I'm not sure what else to try.

jduthon replied:
Tried it myself to see.

We apparently can't do my first idea, as it's not allowed to override a global 😅 

Though the second idea does work:

```typescript
type MockSendBeacon = SpyInstance<Parameters<Navigator["sendBeacon"]>, ReturnType<Navigator["sendBeacon"]>>;
const sendBeaconAsMock = navigator.sendBeacon as unknown as MockSendBeacon;
const actualPayload = JSON.parse(sendBeaconAsMock.mock.calls[0][1]) as IPayload;
```

(ps: this still complains that what we are trying to pass to `JSON.parse` might not be a string as it is typed as `data?: BodyInit | null | undefined`, this sounds like maybe something that should be double checked then whether it's alright to expect it to be a `string` given it shouldn't have such requirement.

mangs replied:
Awesome! This works great, I added `as string` to the parameter being passed to `JSON.parse()` and now zero errors. I still think Vitest should handle this correctly, but thank you again for helping me get to the bottom of this!

mangs reviewed (commented):
Some context comments below, plus there's one weird issue that's not in the code at all:

when I run `tsc`, I get 17 errors targeting Vite-related files in `node_modules` that I can't figure out how to get rid of. I'm using `includes` in `tsconfig.json` so I'm only explicitly including the files I need, but even if I add `exclude: ["node_modules"]` I still have the same problem. Any ideas here would be great because this is my biggest blocker from completion at the moment.

mangs commented:
Down to fixing the last error of spied functions not having `.mock.calls` available on them, so I've reached out to the community for help: https://discord.com/channels/917386801235247114/918057998914568213/1061037463986000024

mangs commented on .github/workflows/publishWorkflow.yml:
```diff
@@ -19,7 +19,8 @@ jobs:
-      - run: npm run build:only
+      - run: npm run build:vite
+      - run: npm run build:types
       # - run: npm publish
       #   env:
       #     NODE_AUTH_TOKEN: ${{ github.token }}
```
NPM publish will get enabled in the next PR. I have to talk to infra about what's involved in pushing code to NPM proper using Babbel's `@babbel` scope

jduthon replied:
Good point, should we rename `@lessonnine/rollbar-client.js` to `@babbel/rollbar-client.js` perhaps?
As I think `lessonnine` makes sense for our internal packages, but for external it might make more sense to go with `babbel`?

mangs replied:
That's exactly right, I've already changed the README to reflect that

jduthon replied:
And I think we also need to update the [package.json](https://github.com/babbel/rollbar-client.js/pull/5/files/a16e6e8e6062eff5c8b5b0bc286d9b2118a7d987#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R2)

mangs commented on config/eslint/eslintConfig.json:
```diff
@@ -1,40 +1,42 @@
 {
   "root": true,
-  "extends": ["./config/eslint/eslintBaseConfig.json", "eslint-config-prettier"],
+  "extends": [
```
Most of the ESLint changes in this PR are direct copy/paste from the ESLint library. Once that gets open sourced, I can remove all the copied code.

mangs commented on config/vite/vite.config.mts:
```diff
@@ -0,0 +1,29 @@
+      fileName: 'RollbarClient',
+      formats: ['es'],
+    },
+    minify: 'esbuild',
```
`esbuild` is the underlying tool written in Go that makes Vite so fast. I'm leveraging it as much as possible to speed up builds and minifies.

mangs commented on config/vite/vite.config.mts:
```diff
@@ -0,0 +1,29 @@
+      formats: ['es'],
+    },
+    minify: 'esbuild',
+    modulePreload: false,
```
This would defeat the purpose of the library

mangs commented on config/vite/vite.config.mts:
```diff
@@ -0,0 +1,29 @@
+    sourcemap: true,
+    target: 'modules',
+  },
+  test: {
+    clearMocks: true,
+    coverage: {
+      provider: 'c8',
+      reportsDirectory: './vitestCodeCoverage',
+    },
+    environment: 'happy-dom',
+  },
```
This is the Vitest-specific section (`test`) of the configuration. Look how easy this is to setup for building and unit tests in one file! And this is in "library mode" which is the less focussed-on case. App builds should be even more straight forward to setup than this.

mangs commented on package.json:
```diff
@@ -11,68 +11,54 @@
   ],
   "exports": {
-    "import": "./dist/RollbarClient.min.mjs",
-    "require": "./dist/RollbarClient.min.cjs"
```
no more CommonJS support, bundlers should handle this fine

mangs commented on package.json:
```diff
@@ -11,68 +11,54 @@
-    "test:watch": "npm run test -- --watch",
-    "update:package-versions": "./scripts/npm/updatePackageVersions.mjs --ignore-packages=rollup && npm run --silent reinstall",
+    "test": "npm-run-all --silent validate:environment test:unit",
+    "test:unit": "vitest --config config/vite/vite.config.mts --coverage",
```
One caveat here is that tests will not fail if less than 100% line coverage is achieved. I'll have to look into adding this back in once Vitest supports this. The underlying tool `c8` that generates coverage reports is not the blocker here, so it should only be a matter of time. Maybe a good excuse for me to contribute a change upstream to Vitest.

mangs commented on src/RollbarClient.test.mts:
```diff
@@ -1,23 +1,27 @@
+// External Imports
+import { beforeAll, describe, expect, test, vi } from 'vitest';
```
So much nicer than magically-appearing globals

mangs reviewed (commented):
This is ready to review. For future use of Vitest in particular, pay attention to how little my unit tests had to change to convert from Jest to Vitest, and that was despite ALSO converting from JS to TS in this same PR.

jduthon commented on tsconfig.json:
```diff
@@ -0,0 +1,30 @@
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "outDir": "./dist",
+    "removeComments": true,
```
Minor but I think it might make sense to keep comments given this is only outputting types, I think they can once in a while bring value to the types as documentation.

mangs replied:
that's a good point

jduthon reviewed (commented):
Looking good overall, great job!

The vite setup is quite interesting and vitest has more compatibility with jest's API then I would have expected :O 

I left a few comments on things that could perhaps be improved, nothing too too drastic I reckon (and nothing super blocking I think, maybe a few of the castings questions and the parameters of the log function are interesting ones to resolve for sure).

mangs commented on src/RollbarClient.mts:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  onErrorDefault = (errorEvent: ErrorEvent) => {
+    this.log('warning', 'Unhandled error occurred', errorEvent.error as Error).catch(() => {});
```
Because `this.log()` is an async function and ESLint wants the promise handled. Since we don't care about the return value (e.g. `sendBeacon()`), I added `.catch()` to handle the error case and stop ESLint complaining.

jduthon replied:
I see.
I am not a big fan in general of swallowing errors without a warning, the error this would report could eventually be interesting during the development phase (I agree that in production, if it's not going to go to your error reporter, it's relatively useless, except perhaps if the developer is the one currently checking the problem 😅).

Maybe there could be an option you can pass to the library something like `silentlyIgnoreNetworkReportingErrors` that one could set accordingly to their wishes?
(I think it wouldn't be useful a lot, but at least it would make the behavior a little more explicit/configurable)

mangs replied:
I've thought about this more, and I've changed this to await the log call so that any exceptions that get thrown will occur at the call site; thank you.

The reason I'm not handling the exceptions is that the only possible exception would come from a failing `fetch` call, and this is an error reporting library wherein the `fetch` call is the main feature, so now if that happens the user will see the exception thrown in their console and know where it came from whereas before it would have been swallowed like you said.

jduthon replied:
Yeah makes sense to me!

mangs replied:
A further adaptation is that I removed the `await` to keep the event listeners non-`async` and instead I'm returning the `this.log()` calls which will have the same effect. There are edge case issues with making event listeners async, so I'm being a bit defensive here because I don't know how end users will use this library. Better safe than sorry.

mangs commented on src/RollbarClientSubmitter.mts:
```diff
@@ -94,29 +110,23 @@ function submitOccurrence(url, payload) {
 class RollbarClientSubmitter {
-  constructor(configurationOptions) {
+  configuration: IConfigurationOptions;
+  errorHistory: Array<object>; // TODO: add more specific object shape
```
For later

mangs commented on src/RollbarClientSubmitter.mts:
```diff
@@ -94,29 +110,23 @@ function submitOccurrence(url, payload) {
     keepalive: true,
     method: 'POST',
-  });
+  }).catch(() => {});
```
I was thinking that since this is an error reporting library, and if connectivity isn't working, there's nothing that can really be done. I suppose adding a console log would be helpful, but it would never reach a reporting service. What do you suggest I do here?

jduthon replied:
I mostly answered with the comment above, I think you are right in most production instances the error is unlikely to be useful, but silencing errors implicitly can also be a dangerous path.
I think we can probably mostly have the discussion up there, I added my arguments but I don't have a super strong opinion as I think the usefulness of those errors would be a rare case.

mangs replied:
Handled; see above

jduthon replied:
The `catch` still seems to be here, and if I have understood correctly the chain of actions within the library, this is the deepest in the call stack, hence catching here will swallow all errors, or am I missing something?

mangs replied:
No you're right. I missed the most important one! 🙃

Thank you for noticing. All fixed now. I had to refactor a bunch of tests because I had to `await fetch` and thus make the containing function `async`, and that's been done too.

mangs commented on src/RollbarClientSubmitter.mts:
```diff
@@ -149,29 +159,30 @@ class RollbarClientSubmitter {
     } = configuration;
-    const title = `${isBrowserSupported ? '' : browserUnsupportedTitlePrefix}${titleText}`;
+    const title = `${
+      isBrowserSupported ? '' : (browserUnsupportedTitlePrefix as string)
```
There is a default in `configurationDefaults` on line 14, so this should never be `undefined`. Is there a more elegant way to handle this without a cast?

jduthon replied:
This means that `this.configuration` isn't typed correctly then, as in the constructor you merge it with the default, the type that `this.configuration` holds shouldn't have `browserUnsupportedTitlePrefix` as an optional anymore.

You could write its type like this to make it accurate:
```
configuration: IConfigurationOptions & Required<Pick<IConfigurationOptions, keyof typeof configurationDefaults >>;
```

It's probably even possible to get away with:
```
configuration: IConfigurationOptions & typeof configurationDefaults;
```

mangs replied:
Omg this is brilliant, thank you!

mangs commented on src/types.mts:
```diff
@@ -0,0 +1,112 @@
+};
+
+type TSubmitterParameters = [
+  string,
```
Great idea, I've added the log levels as the first argument to the type. When I try instantiating a new `RollbarClient` instance and calling `client.log()`, the first argument gets auto-filled with the five possible log levels. Is that what you meant?

jduthon replied:
So that's what I meant with the very first part of this comment, but this still doesn't address the second part.

I believe at the moment, when you type `client.log()`, your autocomplete will give you the type of each parameter the function takes, but they will be unnamed, so it's a bit hard as a consumer to know what should go into it (I must read the docs).

If you type the function regularly with each parameter named, then the editor will also give consumers of the library the name of each parameter in addition to their types, which should help quite a bit.

mangs replied:
Oh i got you now, that IS a good idea. I've added it and indexed `TSubmitterParameters` for each parameter so they stay in sync. Thank you for the suggestion!

mangs commented on src/types.mts:
```diff
@@ -0,0 +1,112 @@
+  locationInfo?: object;
+  setContext: () => string;
+  shouldIgnoreOccurrence: (payload: IPayload, configuration: IConfigurationInternal) => boolean;
+  userInfo?: object;
```
This is a good observation, but it makes comparing the two interfaces a lot harder. `IConfigurationInternal` has many of the same keys as `IConfigurationOptions` but not all, many are required but not all. I figured it was easier to read by writing it manually like this than a series of `Pick`, `Required`, `Omit`, and `Partial` types mixed together.

jduthon replied:
I think the comparing argument is debatable.
On the one hand they are easier to read one by one, on the other comparing them right now requires careful scanning of each item to spot the difference, whereas something like:
```typescript
type TConfigurationInternal = 
  Omit<TConfigurationOptions, 'onUnhandledError'> &
  Required<Pick<TConfigurationOptions, 'access_token' | 'browsersSupportedRegex'>>
```
makes the difference between each much more explicit (but potentially a bit harder to read point-blank, agreed on that part).
I think the main risk with not aligning them would be that modifying one without modifying the other could be an error, but hopefully if it's an important error something else in the code should catch it (as long as there is no cast).
So resolving as you made the decision.

mangs commented on src/types.mts:
```diff
@@ -0,0 +1,112 @@
+
+type TCustomConfiguration = Omit<IConfigurationInternal, 'accessToken'> & {
+  onUnhandledError?: string;
+  onUnhandledPromiseRejection?: string;
```
Nevermind, the `hasConfigurationInPayload` test breaks when you remove this

mangs replied:
Not sure why `TSerializedObject` isn't picking those up properly, seems like it should. But these are needed to eliminate type errors.

jduthon replied:
It looks like `TSerializedObject` is wrong indeed 🤔 

Somehow [this version](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYgrgOwMbAJYHsEEEDOBlYAJ1QQHMAeAFQD4oBeKSqCAD2AgQBMdZEUMEUAPxQcREqSgAuRgG4AUKEhQAShFIBRFmFwFiZKrQZNW7LjzWbtw0eLLS58+SXaEAZgEMk0GB4DWEADyAEYAVhAoUADe8lBxUOxiMp4ANjjQAD68yGiYCvEJEGIATDLwOQL58YnAAMzJHmmZqupaYFVxNQAsMmL6pB2FYgCsMghwALbBEIRQWX0Sc9n8mEuWbQoAvgqK4NCUeDOojagAXhCcIeEoVMxsHNxQ6GERwEbRsVAA2gEgUCRQX7oNyMAC6MnW2l0dgo5RW2HwMKoPwgIFB1GoWx2u2Uh2IJ3OnF8ASur3ojDxxxSZwupJuxKCLxQmKAA) works, it looks pretty similar but I am guessing having an intermediate type makes the `extends` apply on each member of the union instead of on the union as a whole 🤷‍♂️

mangs commented on src/types.mts:
```diff
@@ -0,0 +1,112 @@
+  userInfo?: object;
+}
+
+interface IGenericObjectIndexSignature {
```
done, ive removed it

jduthon commented on src/RollbarClientSubmitter.mts:
```diff
@@ -183,16 +194,18 @@ class RollbarClientSubmitter {
+        context: setContext ? setContext() : configurationDefaults.setContext(),
         custom: {
-          isBrowserSupported,
+          isBrowserSupported: isBrowserSupported ?? true,
```
I see it is documented, looks good then! 👍

mangs replied:
I'll try to make this more clear in my doc revisions

jduthon commented on src/RollbarClientSubmitter.mts:
```diff
@@ -241,18 +254,18 @@ class RollbarClientSubmitter {
 
     // Submit occurrence to Rollbar
-    submitOccurrence(apiUrl, payload);
+    submitOccurrence(apiUrl as string, payload);
```
I think if we fix the type of `configuration` as per a comment above, the cast shouldn't be required anymore (nor should a bunch of the defaulting to the `configurationDefaults.property`)

mangs replied:
you were right, i removed this

jduthon commented on src/RollbarClientSubmitter.mts:
```diff
@@ -241,18 +254,18 @@ class RollbarClientSubmitter {
-      typeof shouldIgnoreOccurrence === 'function' &&
-      shouldIgnoreOccurrence(payload, this.configuration)
+      shouldIgnoreOccurrence &&
+      shouldIgnoreOccurrence(payload, this.configuration as IConfigurationInternal)
```
Yeah I think that's another sign currently `this.configuration` isn't typed correctly.

mangs commented:
@jduthon aside from the comment you made about typing the class and my response, I've addressed all your comments. Thank you so much for your valuable insight! Seeming like we're close to the finish line 🏁

jduthon commented on src/types.mts:
```diff
@@ -0,0 +1,91 @@
+
+type TConfigurationSerialized<T extends object> = {
+  // eslint-disable-next-line @typescript-eslint/ban-types -- any function can be expected here
+  [key in keyof T]: Exclude<T[key], undefined | false> extends RegExp | Function ? string : T[key];
```
This will be ever so slightly worst than what I had provided, because it will transform e.g. `RegExp | false` to `string`, while ideally it should be `string | false` (or perhaps an optional string)

mangs replied:
This took awhile to figure out why your code was causing errors, and I figured it out.

First off, addressing your comment, there is no field expecting `RegExp | false`, I know that must have been more confusing with the `IConfigurationOptions` and `IConfigurationInternal`; now only the prior exists. For that use case, either the expected data types are used, or a `string` is expected in the `custom.configuration` object for `RegExp` and `Function` objects so they can be serialized with `toString()`.

The reason the very clever type code you wrote was causing errors on a few fields (namely the `RegExp` and `Function` fields) is because (1) those fields were optional and (2) I have the `tsconfig` option `exactOptionalPropertyTypes` enabled which causes optional fields to have `undefined` added implicitly as part of a union type. That's why I had to explicitly `Exclude<>` `undefined`; `Required<>` did not do the trick here unexpectedly.

Given all this, what are your thoughts?

jduthon replied:
On the code I proposed (maybe not the very first one, but the improved version I had put in a comment), I don't think `undefined` should be any problem (I tried it out in the example below too and things seem to work how I'd expect them to).

Checking also your current solution for `undefined`, it is quite interesting the results 👀 
`Function | undefined` is transformed to `string`, which isn't ideal.
On the other hand, optionality seems to work, so that `test?: Function` is transformed into `test?: string`, I am guessing `[key in keyof T]` keeps the optionality, hence why it seems to work for what is here.

It's slightly less complete than the other implementation as it won't deal properly with a field that can be a RegExp/Function or something else (e.g, `Function | number`), cf [this example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYgrgOwMbAJYHsEEEDOBlYAJ1QQHMAeAFQD4oBeKSqCAD2AgQBMdZEUMEUAPxQcREqSgAuRgG4AUKEhQAShFIBRFmFwFiZKrQZNW7LjzWbtw0eLLS58p-JLtCAMwCGSaDE8BrCAB5ACMAKwgUKABveSh4qHYxIRl4ZDRMBQTEiDEAJlS+DIQshKTgAGYZSy0wKAAfKEROCHcSCE5S+PKAFhkxfVIunLEAVhSoBDgAWxCIQgbbQcW0-kxFmu0FAF8FRXBoSjx51E8AG1QALw7QiJQqZjYObih0cMjgIxi4qABtQJAUBIUAB6HcjAAutV1LVdHYKKtinDBlR-hAQBDqNQdnslIcAMKYNqkOCETzFY7Ec5XDoPUzPHhvO6fejfeIAenZzBwFwQwAAtJxUDhPCEzhB+QhTPzedAAAJ4nBIYhgAW5XnAdkhTwIfmKqD8-lQHWA9xFARQJA6qBzR6QFAdKAAC3mEB+aMBwNB4MoUKgWiQZzgLVRAIhABomlxWu1OIsvGccBBaPTzKoYdZGoiLSIBhIHJQPRCcU5OVAAOqoYBOxJO4WvKWR85nWvQfWeQjQJDoQidlBnQFEHU4dw96YdSM4dC18kjYCjKCkCDAHisbzAAetqDEUhO4BQABu5zguX2ykppwu104fkCtw+rKOJ2p1-v91vwXeKGx8jLlertb1pgECRuUC4AO6YAA5PuS77tW0A7nuh7HrkNqRJ4cBJlA4HQB20CYJuwAdnBaD2AhohwCESb7gABs0MZSnGjQJkmtFntAF4vh0H5vsA-5OgACioACS0xgC2xiEggxKkuSAhcVetK8V+nyyEAA), but you don't have anything like that at the moment, so not critical.

mangs replied:
Asking question via Slack

jduthon commented:
@mangs
Thanks a lot for taking care of the comments 🙌 

I answered to the comment about the class/prototype thingy, but given it's small and in a test file, it's not a blocker.

What about [this one](https://github.com/babbel/rollbar-client.js/pull/5#discussion_r1073548166) though?
I think fixing this would make the library easier to use if I understood things correctly and I think it could be worth addressing.

The last comment I added is minor and not a blocker, as ultimately those things are going to a logger, so whether a field is `string | undefined` or `string` shouldn't matter too much for the logging purposes.

mangs commented:
@jduthon Final pass I hope...I hope I didn't jinx myself. Thanks for the great feedback. This is looking a lot better than a few weeks ago. Let's see if we can get this merged Monday. =D

jduthon reviewed (approved):
Thanks a lot for addressing all the feedback! 👏 

Great work! 🚀 

I realized we also forgot of [this comment](https://github.com/babbel/rollbar-client.js/pull/5#discussion_r1073581919) but given it's in a test file I don't think it's a blocker.
I also answered to the comment on the serialized thingy, but given the current version works with the code, my proposal is only an improvement, so we can go without it.

mangs commented:
@jduthon not sure how i missed that comment, I just addressed it
